elasticsearch/src/main/scala/org/elasticsearch/spark/sql/ElasticsearchXDRelation.scala

```diff
@@ -37,9 +88,256 @@ import org.apache.spark.sql.{Row, SQLContext}
+      // we could use regexp filter however it's a bit overkill and there are plenty of chars to escape
+      // s"""{"regexp":{"$attribute":"$value.*"}}"""
+      // as an alternative we could use a query string but still, the analyzed / non-analyzed is there as the DSL is slightly more complicated
+      // s"""{"query":{"query_string":{"default_field":"$attribute","query":"$value*"}}}"""
+      // instead wildcard query is used, with the value lowercased (to match analyzed fields)
+
+      case f: Product if isClass(f, "org.apache.spark.sql.sources.StringStartsWith") => {
+        var arg = f.productElement(1).toString()
+        if (!strictPushDown) {
+          arg = arg.toLowerCase(Locale.ROOT)
+        }
+        s"""{"query":{"wildcard":{"${f.productElement(0)}":"$arg*"}}}"""
+      }
+
+      case f: Product if isClass(f, "org.apache.spark.sql.sources.StringEndsWith") => {
+        var arg = f.productElement(1).toString()
+        if (!strictPushDown) {
+          arg = arg.toLowerCase(Locale.ROOT)
+        }
+        s"""{"query":{"wildcard":{"${f.productElement(0)}":"*$arg"}}}"""
+      }
+
+      case f: Product if isClass(f, "org.apache.spark.sql.sources.StringContains") => {
+        var arg = f.productElement(1).toString()
+        if (!strictPushDown) {
+          arg = arg.toLowerCase(Locale.ROOT)
+        }
+        s"""{"query":{"wildcard":{"${f.productElement(0)}":"*$arg*"}}}"""
+      }
+
+      // the filters below are available only from Spark 1.5.0
+
+      case f: Product if isClass(f, "org.apache.spark.sql.sources.EqualNullSafe") => {
+        var arg = extract(f.productElement(1))
+        if (strictPushDown)
+          s"""{"term":{"${f.productElement(0)}":$arg}}"""
+        else
+          s"""{"query":{"match":{"${f.productElement(0)}":$arg}}}"""
+      }
+
+      case _ => ""
+    }
+  }
+
+    def createDSLFromFilters(filters: Array[Filter], strictPushDown: Boolean) = {
+      filters.map(filter => translateFilter(filter, strictPushDown)).filter(query => StringUtils.hasText(query))
+    }
+
+  }
+
+  // PrunedFilteredScan
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]) = {
+    import CopiedCode._
+    val paramWithScan = mutable.LinkedHashMap[String, String]() ++ parameters
+    paramWithScan += (InternalConfigurationOptions.INTERNAL_ES_TARGET_FIELDS ->
+      StringUtils.concatenate(requiredColumns.asInstanceOf[Array[Object]], StringUtils.DEFAULT_DELIMITER))
+
+    // scroll fields only apply to source fields; handle metadata separately
+    if (cfg.getReadMetadata) {
+      val metadata = cfg.getReadMetadataField
+      // if metadata is not selected, don't ask for it
+      if (!requiredColumns.contains(metadata)) {
+        paramWithScan += (ConfigurationOptions.ES_READ_METADATA -> false.toString())
+      }
+    }
+
+    if (filters != null && filters.size > 0) {
```

Comment:
Absolutely but this is code coming from the connector, we agreed that copy-pasted code should remain untouched for maintenance purposes.